accounts/abi/bind/dep_tree.go

```diff
@@ -0,0 +1,140 @@
+package bind
+
+import (
+	"encoding/hex"
+	"fmt"
+	"github.com/ethereum/go-ethereum/common"
+	"github.com/ethereum/go-ethereum/core/types"
+	"maps"
+	"strings"
+)
+
+// DeploymentParams represents parameters needed to deploy a
+// set of contracts.  It takes an optional override
+// list to specify contracts/libraries that have already been deployed on-chain.
+type DeploymentParams struct {
+	contracts []*MetaData
+	// map of solidity library pattern to constructor input.
+	inputs map[string][]byte
+	// Overrides is an optional map of pattern to deployment address.
+	// Contracts/libraries that refer to dependencies in the override
+	// set are linked to the provided address (an already-deployed contract).
+	overrides map[string]common.Address
+}
+
+// NewDeploymentParams instantiates a DeploymentParams instance.
+func NewDeploymentParams(contracts []*MetaData, inputs map[string][]byte, overrides map[string]common.Address) *DeploymentParams {
+	return &DeploymentParams{
+		contracts,
+		inputs,
+		overrides,
+	}
+}
+
+// DeploymentResult encapsulates information about the result of the deployment
+// of a set of contracts: the pending deployment transactions, and the addresses
+// where the contracts will be deployed at.
+type DeploymentResult struct {
+	// map of contract library pattern -> deploy transaction
+	Txs map[string]*types.Transaction
+	// map of contract library pattern -> deployed address
+	Addrs map[string]common.Address
+}
+
+// Accumulate merges `other` into `d`
+func (d *DeploymentResult) Accumulate(other *DeploymentResult) {
+	maps.Copy(d.Txs, other.Txs)
+	maps.Copy(d.Addrs, other.Addrs)
+}
+
+// DeployFn deploys a contract given a deployer and optional input.  It returns
+// the address and a pending transaction, or an error if the deployment failed.
+type DeployFn func(input, deployer []byte) (common.Address, *types.Transaction, error)
+
+// depTreeDeployer is responsible for taking a dependency, deploying-and-linking its components in the proper
+// order.  A depTreeDeployer cannot be used after calling LinkAndDeploy other than to retrieve the deployment result.
+type depTreeDeployer struct {
+	overrideAddrs map[string]common.Address
+	deployedAddrs map[string]common.Address
```

Comment:
I think you can drop `overrideAddrs`, and shove everything into `deployedAddrs`. The deployerTxs can be used if the user wants to know which happened now and which happened previously

Reply:
WDYT about filtering the overrides out of `deployedAddrs` at the end of `LinkAndDeploy` based on their non-presence in the map of deployed txs?

Reply:
Something like that, yes. 
Here are some suggested changes: https://github.com/holiman/go-ethereum/commit/4ad21ba34e1f30c39bbb127095e95cd33592f447 
I wanted to try it out in code to see if it worked. Seems to, so far

Reply:
^ as you can see in that commit, I don't see any need for `Accumulate` to exist? It's accumulated in the deployer